src/main/kotlin/com/autonomousapps/tasks/FilterAdviceTask.kt

```diff
@@ -177,13 +196,57 @@ abstract class FilterAdviceTask @Inject constructor(
       output.bufferWriteJson(filteredAdvice)
     }
 
-    private fun Sequence<Advice>.filterOf(behavior: Behavior, predicate: (Advice) -> Boolean): Sequence<Advice> {
+    // Don't delete yet.
+//    private fun Sequence<Advice>.filterOf(behavior: Behavior, predicate: (Advice) -> Boolean): Sequence<Advice> {
+//      return filterNot { advice ->
+//        predicate(advice) && (
+//          behavior is Ignore
+//            || behavior.filter.contains(advice.coordinates.identifier)
+//            || behavior.filter.contains(advice.coordinates.gav())
+//          )
+//      }
+//    }
+
+    private fun Sequence<Advice>.filterOf(
+      behaviorSpec: Pair<Behavior, List<Behavior>>,
+      predicate: (Advice) -> Boolean
+    ): Sequence<Advice> {
+      val globalBehavior = behaviorSpec.first
+      val sourceSetsBehavior = behaviorSpec.second
+
+      val byGlobal: (Advice) -> Boolean = { a ->
+        globalBehavior is Ignore
+          || globalBehavior.filter.contains(a.coordinates.identifier)
+          || globalBehavior.filter.contains(a.coordinates.gav())
+      }
+
+      val bySourceSets: (Advice) -> Boolean = { a ->
+        // These are the source sets represented in this advice. Might be empty if it is for the main source set.
+        val adviceSourceSets = supportedSourceSets
+          .map { it.lowercase() }
+          .filter { s ->
+            val from = a.fromConfiguration?.lowercase()?.startsWith(s) == true
+            val to = a.toConfiguration?.lowercase()?.startsWith(s) == true
+            from || to
```

Comment:
Maybe I miss something, but does this work correctly for the `main` source set? Because if `s == ""` it matches everything, also things that **are not in main**.

If you want this to be 100% accurate I think you would have to split the sourceSetName out of the configuration name, by also looking for all the possible suffixes ("Implementation", "Api", "CompileOnly", ...):
1. Match one of the suffixes
2. Strip the suffix
3. The result is the _source set name_ or if `""` it is `main`

I did that here. If you ignore the Java Modules Directive part:

- Have a list of the possible suffixes: https://github.com/gradlex-org/java-module-dependencies/blob/main/src/main/java/org/gradlex/javamodule/dependencies/tasks/ModuleDirectivesScopeCheck.java#L48-L54
- Compute the source set name by first determining which of the the suffixes ("Scope Configuration") is used and then strip that to get the source set name https://github.com/gradlex-org/java-module-dependencies/blob/main/src/main/java/org/gradlex/javamodule/dependencies/tasks/ModuleDirectivesScopeCheck.java#L139-L154

Reply:
How could `s == ""`, though? `s` is an element of the collection `supportedSourceSets`, which in a "normal" case might have values like "main" and "test." The comment above that says "Might be empty if it is for the main source set" means that the _result_ of the map+filter might be empty. (is that comment what you're referring to?)

I considered your approach that you linked, but for some reason having a concrete list of scopes was bothering me :) Any chance Gradle will add yet another possibility? That said, I'll continue changing the implementation to match your preference because it would probably be more readable.

Reply:
Sorry then I misunderstood. I thought for the "main" source set that list would contain the empty string. But if it contains "main", does it work then? Because the configurations do not start with "main".

Anyway, if this works and I just miss something, then you can also keep the approach you have.

Regarding the list of "scopes": I don't see how that would ever change again. All combinations that makes sense in Java are now covered (since the addition of compileOnlyApi).

Reply:
Maybe this list already exists somewhere in the Plugin code? In the end, these are also the configurations/scopes for which advices are created in the first place.

Reply:
I just pushed a commit that follows your advice and I think the code is much cleaner and easier to follow. And the tests still pass! So thanks :)